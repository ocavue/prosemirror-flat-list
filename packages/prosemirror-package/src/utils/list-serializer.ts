import {
  DOMOutputSpec,
  DOMSerializer,
  Fragment,
  Node as ProsemirrorNode,
  Schema,
} from 'prosemirror-model'
import { listToDOM } from '../schema/to-dom'

/** @public */
export class ListDOMSerializer extends DOMSerializer {
  static nodesFromSchema(schema: Schema): {
    [node: string]: (node: ProsemirrorNode) => DOMOutputSpec
  } {
    const nodes = DOMSerializer.nodesFromSchema(schema)
    return { ...nodes, list: (node) => listToDOM(node, true) }
  }

  serializeFragment(
    fragment: Fragment,
    options?: { document?: Document },
    target?: HTMLElement | DocumentFragment,
  ): HTMLElement | DocumentFragment {
    const dom = super.serializeFragment(fragment, options, target)
    return joinListElements(dom)
  }
}

/**
 * Merge adjacent <ul> elements or adjacent <ol> elements into a single list element.
 *
 * @internal
 */
export function joinListElements<T extends Element | DocumentFragment>(
  parent: T,
): T {
  for (let i = 0; i < parent.childNodes.length; i++) {
    const child = parent.children.item(i)
    if (!child) continue

    if (child.tagName === 'UL' || child.tagName === 'OL') {
      let next: Element | null = null

      while (
        ((next = child.nextElementSibling), next?.tagName === child.tagName)
      ) {
        child.append(...Array.from(next.children))
        next.remove()
      }
    }
    joinListElements(child)
  }
  return parent
}
