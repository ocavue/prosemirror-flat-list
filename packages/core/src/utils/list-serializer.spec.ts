import { describe, expect, it } from 'vitest'

import { setupTestingEditor } from '../../test/setup-editor'

import { ListDOMSerializer } from './list-serializer'

describe('ListDOMSerializer', () => {
  const {
    add,
    doc,
    p,
    bulletList,
    orderedList,
    uncheckedTaskList: taskList,
    expandedToggleList: toggleList,
    schema,
  } = setupTestingEditor()

  let editor: ReturnType<typeof add>

  it('can serialize list nodes into <ul>', () => {
    editor = add(doc(bulletList(p('A')), bulletList(p('B'))))

    const serializer = ListDOMSerializer.fromSchema(schema)

    const serialized = serializer.serializeFragment(editor.state.doc.content)
    expect(serialized.querySelectorAll('ul').length).toBe(1)
    expect(serialized.querySelectorAll('ol').length).toBe(0)
    expect(serialized.querySelectorAll('ul > li').length).toBe(2)
    expect(serialized).toMatchSnapshot()
  })

  it('can serialize list nodes into <ol>', () => {
    editor = add(doc(orderedList(p('A')), orderedList(p('B'))))

    const serializer = ListDOMSerializer.fromSchema(schema)

    const serialized = serializer.serializeFragment(editor.state.doc.content)

    expect(serialized.querySelectorAll('ul').length).toBe(0)
    expect(serialized.querySelectorAll('ol').length).toBe(1)
    expect(serialized.querySelectorAll('ol > li').length).toBe(2)
    expect(serialized).toMatchSnapshot()
  })

  it('can serialize list nodes with different types into a single <ul>', () => {
    editor = add(
      doc(
        bulletList(p('A')),
        taskList(p('B')),
        toggleList(p('C')),

        orderedList(p('D')),

        bulletList(p('D')),
        taskList(p('E')),
        toggleList(p('D')),
      ),
    )

    const serializer = ListDOMSerializer.fromSchema(schema)

    const serialized = serializer.serializeFragment(editor.state.doc.content)

    expect(serialized.querySelectorAll('ul').length).toBe(2)
    expect(serialized.querySelectorAll('ol').length).toBe(1)
    expect(serialized.querySelectorAll('ul > li').length).toBe(6)
    expect(serialized.querySelectorAll('ol > li').length).toBe(1)
    expect(serialized).toMatchSnapshot()
  })

  it('can serialize nested list node ', () => {
    editor = add(
      doc(
        bulletList(p('A'), orderedList(p('B')), orderedList(p('C'))),
        bulletList(p('D'), orderedList(p('E')), orderedList(p('F'))),
      ),
    )

    const serializer = ListDOMSerializer.fromSchema(schema)

    const serialized = serializer.serializeFragment(editor.state.doc.content)

    expect(serialized.querySelectorAll('ul').length).toBe(1)
    expect(serialized.querySelectorAll('ol').length).toBe(2)
    expect(serialized.querySelectorAll('ul > li').length).toBe(2)
    expect(serialized.querySelectorAll('ol > li').length).toBe(4)
    expect(serialized).toMatchSnapshot()
  })
})
