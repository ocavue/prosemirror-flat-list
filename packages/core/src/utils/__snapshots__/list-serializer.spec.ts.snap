// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ListDOMSerializer > can serialize list nodes into <ol> 1`] = `
<DocumentFragment>
  <ol>
    <li
      class="prosemirror-flat-list"
      data-list-kind="ordered"
    >
      <p>
        A
      </p>
    </li>
    <li
      class="prosemirror-flat-list"
      data-list-kind="ordered"
    >
      <p>
        B
      </p>
    </li>
  </ol>
</DocumentFragment>
`;

exports[`ListDOMSerializer > can serialize list nodes into <ul> 1`] = `
<DocumentFragment>
  <ul>
    <li
      class="prosemirror-flat-list"
      data-list-kind="bullet"
    >
      <p>
        A
      </p>
    </li>
    <li
      class="prosemirror-flat-list"
      data-list-kind="bullet"
    >
      <p>
        B
      </p>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`ListDOMSerializer > can serialize list nodes with different types into a single <ul> 1`] = `
<DocumentFragment>
  <ul>
    <li
      class="prosemirror-flat-list"
      data-list-kind="bullet"
    >
      <p>
        A
      </p>
    </li>
    <li
      class="prosemirror-flat-list"
      data-list-kind="task"
    >
      <p>
        B
      </p>
    </li>
    <li
      class="prosemirror-flat-list"
      data-list-kind="toggle"
    >
      <p>
        C
      </p>
    </li>
  </ul>
  <ol>
    <li
      class="prosemirror-flat-list"
      data-list-kind="ordered"
    >
      <p>
        D
      </p>
    </li>
  </ol>
  <ul>
    <li
      class="prosemirror-flat-list"
      data-list-kind="bullet"
    >
      <p>
        D
      </p>
    </li>
    <li
      class="prosemirror-flat-list"
      data-list-kind="task"
    >
      <p>
        E
      </p>
    </li>
    <li
      class="prosemirror-flat-list"
      data-list-kind="toggle"
    >
      <p>
        D
      </p>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`ListDOMSerializer > can serialize nested list node  1`] = `
<DocumentFragment>
  <ul>
    <li
      class="prosemirror-flat-list"
      data-list-collapsable=""
      data-list-kind="bullet"
    >
      <p>
        A
      </p>
      <ol>
        <li
          class="prosemirror-flat-list"
          data-list-kind="ordered"
        >
          <p>
            B
          </p>
        </li>
        <li
          class="prosemirror-flat-list"
          data-list-kind="ordered"
        >
          <p>
            C
          </p>
        </li>
      </ol>
    </li>
    <li
      class="prosemirror-flat-list"
      data-list-collapsable=""
      data-list-kind="bullet"
    >
      <p>
        D
      </p>
      <ol>
        <li
          class="prosemirror-flat-list"
          data-list-kind="ordered"
        >
          <p>
            E
          </p>
        </li>
        <li
          class="prosemirror-flat-list"
          data-list-kind="ordered"
        >
          <p>
            F
          </p>
        </li>
      </ol>
    </li>
  </ul>
</DocumentFragment>
`;
